### PR TITLE
server: simplify tenant auto upgrade logic

### LINKED ISSUE #121858: server: tenant auto upgrade may not activate if cluster setting is set
If one of `cluster.auto_upgrade.enabled` or `cluster.preserve_downgrade_option` are set in such a way to stop auto-upgrades from happening at the time the storage cluster finishes upgrading, then tenants will not attempt an auto upgrade when those settings are reset.

This is because the logic that checks if an auto upgrade should be attempted doesn't take those settings into account:

https://github.com/cockroachdb/cockroach/blob/f8f7dcd6d1a0fcc293094dfff351abbe5505f108/pkg/server/tenant_auto_upgrade.go#L59-L61

Internal discussion can be found [on this link](https://cockroachlabs.slack.com/archives/C02HWA24541/p1710965085964679).

Jira issue: CRDB-37589

### PR DESCRIPTION
**Pull request by renatolabs**
This commit fixes a bug in the tenant auto upgrade logic where tenants
will not attempt to auto upgrade if an auto-upgrade related cluster
setting (`cluster.auto_upgrade.enabled` or
`cluster.preserve_downgrade_option`) was active at the time the
storage cluster finished upgrading. In that case, the tenant would
never pick up the fact that it is now able to auto upgrade, and would
remain in the older cluster version (auto upgrading only if the
service is restarted).

The fix here also brings the tenant auto upgrade logic a lot closer to
the system auto upgrade counterpart. We continue to attempt an
auto-upgrade until we know we have upgraded as much as possible (to
the same version as the storage cluster).

Fixes: #121858

Release note: None

### CODE REVIEW

**Comment by renatolabs** [comment]
I think this should be ready for review. The ownership story for this part of the code is unclear. Pinging @ajstorm and @stevendanna as potential reviewers.

**Comment by srosenberg** [comment]
@dt maybe we could enlist your help to get this fix reviewed?

**Comment by stevendanna** [comment]
Apologies I missed this, went from PTO into an on-site which has disrupted my usual PR review process. I'll review this by tomorrow at the latest.

File: pkg/server/tenant_auto_upgrade.go [hunk 1]
```diff
@@ -31,40 +31,41 @@ import (
...
-					firstAttempt = false
-					storageClusterVersion = latestStorageClusterVersion
-					if err := s.startAttemptTenantUpgrade(ctx); err != nil {
+			case <-time.After(loopFrequency):
+				storageClusterVersion := s.settingsWatcher.GetStorageClusterActiveVersion().Version
+
+				// Only attempt to upgrade if the storage cluster is at an Internal version of 0,
+				// which implies that storage is at the "final" version for some release.
+				if storageClusterVersion.Internal == 0 {
+					log.Infof(ctx, "starting auto upgrade attempt")
```
Comment (stevendanna) [inline pkg/server/tenant_auto_upgrade.go:50]: You've been working with this and have a better sense of what is useful, but I wonder if we want to remove this log message as we already have a number of them in startAttemptTenantUpgrade.

**Review by stevendanna — APPROVED** [review]
Thanks for digging into this!

File: pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go [hunk 2]
```diff
@@ -102,140 +117,83 @@ func TestTenantAutoUpgradeRespectsAutoUpgradeEnabledSetting(t *testing.T) {
...
-	// Wait for auto upgrade status to be received by the testing knob.
-	succeedsSoon := 20 * time.Second
-	for {
-		select {
-		case upgradeInfo := <-upgradeInfoCh:
-			if int(server.UpgradeDisabledByConfiguration) == upgradeInfo.Status {
-				return
+	waitForUpgradeInfo := func(expectedVersion roachpb.Version, expectedStatus int) {
+		succeedsSoon := 20 * time.Second
+		if util.RaceEnabled || skip.Stress() {
```
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go]: Added, thanks!

File: pkg/server/tenant_auto_upgrade.go [hunk 3]
```diff
@@ -113,20 +115,23 @@ func (s *SQLServer) startAttemptTenantUpgrade(ctx context.Context) error {
...
-		return err
+		return false, err
 	case UpgradeDisabledByConfiguration:
-		log.Infof(ctx, "auto upgrade is disabled for current version (preserve_downgrade_option): %s", redact.Safe(tenantClusterVersion.Version))
-		return nil
+		log.Infof(ctx, "auto upgrade is disabled for current version (cluster.auto_upgrade.enabled): %s", redact.Safe(tenantClusterVersion.Version))
+		return false, nil
+	case UpgradeDisabledByConfigurationToPreserveDowngrade:
+		log.Infof(ctx, "auto upgrade is disabled for current version (cluster.preserve_downgrade_option): %s", redact.Safe(tenantClusterVersion.Version))
+		return false, nil
```
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/server/tenant_auto_upgrade.go:122]: Yes, that's what happened before. This early validation is closer to what the system tenant auto upgrade logic does, and I think it's good to have them behave similarly.

File: pkg/server/tenant_auto_upgrade.go (hunk 1 above)
Comment (renatolabs) [reply pkg/server/tenant_auto_upgrade.go]: Good point, this is not really needed or useful. Removed.

**Comment by renatolabs** [comment]
TFTR!

bors r=stevendanna

**Pull request merged**